### PR TITLE
Implement #11235, Resourceable and Roleable, Function in allow method

### PR DESCRIPTION
**Pull request by Jurigag**
Add option to add function in allow method in Phalcon\Acl\Adapter\Memory like this:

``` php
$acl->allow('SomeRole','SomeResource','update',function(TestResource $user,TestModel $model){
    return $user->getId() == $model->getUser();
});
```

 Where we can put class implemting Roleable as first argument, and Resourceable as second argument. Then when we do

``` php
$acl->isAllowed($roleable,$resourcable,'update');
```

It will call our defined method in allow if there is any.

### CODE REVIEW

**Comment by Jurigag** [comment]
I dont know why test didnt passed. Its implementing this my interface, for me it looks fine, weird.

File: phalcon/acl/roleable.zep [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * Phalcon\Acl\Roleable
+ *
+ * Interface for classes which could be used in allow method as ROLE, RoleInterace need __construct method so i decided to do this simple interface
+ */
+interface Roleable
+{
+    /**
+     * Returns the role name
+    */
+    public function getRoleName() -> string;
```
Comment (sergeyklay) [inline phalcon/acl/roleable.zep]: https://github.com/phalcon/cphalcon/blob/master/phalcon/acl/roleinterface.zep#L38 is not enough?

File: phalcon/acl/resourceable.zep [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * Phalcon\Acl\Resource
+ *
+ * Interface for classes which could be used in allow method as RESOURCE, ResourceInterface need __construct method so i decided to do this simple interface
+ */
+interface Resourceable
+{
+    /**
+     * Returns the role name
+    */
+    public function getResourceName() -> string;
```
Comment (sergeyklay) [inline phalcon/acl/resourceable.zep]: https://github.com/phalcon/cphalcon/blob/master/phalcon/acl/resourceinterface.zep#L39  is not enough?

Comment (Jurigag) [inline phalcon/acl/resourceable.zep]: Well i wrote in comment exactly why ResourceInterface/RoleInterface is kind of problem. For example if we want to check have a class which extends Model then we cant implement ResourceInterface/RoleInterface beacause Model has his own final construct class.

File: tests/unit/Phalcon/Acl/Adapter/Memory/ACLMemoryTest.php [hunk 3]
```diff
@@ -247,7 +250,7 @@ public function testAclAddResourceExists()
             'Acl\Resource does not exist in Acl',
             function () {
 
-                $acl         = new PhTAclMem();
+                $acl = new PhTAclMem();
```
Comment (sergeyklay) [inline tests/unit/Phalcon/Acl/Adapter/Memory/ACLMemoryTest.php]: Rollback please all the changes are not relevant to your PR explicitly

File: phalcon/acl/adapter/memory.zep [hunk 4]
```diff
@@ -619,7 +666,16 @@ class Memory extends Adapter
...
+		/**
+		 * If we have both objects and funcAccess call funcAccess too
+		 */
+
+		if (resourceObject != null && roleObject != null && funcAccess != null){
+		 	return (haveAccess == Acl::ALLOW) && call_user_func_array(funcAccess,[roleObject,resourceObject]);
+		}
+		else{
+			return (haveAccess == Acl::ALLOW);
+		}
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: else not needed here, just `return ` is enough

File: phalcon/acl/adapter/memory.zep [hunk 5]
```diff
@@ -619,7 +666,16 @@ class Memory extends Adapter
 			return false;
 		}
 
-		return (haveAccess == Acl::ALLOW);
+		/**
+		 * If we have both objects and funcAccess call funcAccess too
+		 */
+
+		if (resourceObject != null && roleObject != null && funcAccess != null){
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: parentheses are not needed

File: phalcon/acl/adapter/memory.zep [hunk 6]
```diff
@@ -495,17 +506,40 @@ class Memory extends Adapter
...
+				let roleObject = roleName;
+				let roleName = roleObject->getRoleName();
+			}
+			else{
+				throw new Exception("Object passed as roleName must implement Roleable");
+			}
+		}
+
+		if typeof resourceName == "object" {
+			if(resourceName instanceof Resourceable){
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: parentheses are not needed

File: phalcon/acl/adapter/memory.zep [hunk 7]
```diff
@@ -495,17 +506,40 @@ class Memory extends Adapter
...
+
+
+
+		if typeof roleName == "object" {
+			if(roleName instanceof Roleable){
+				let roleObject = roleName;
+				let roleName = roleObject->getRoleName();
+			}
+			else{
+				throw new Exception("Object passed as roleName must implement Roleable");
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: just throw without else

File: phalcon/acl/adapter/memory.zep [hunk 8]
```diff
@@ -349,7 +358,7 @@ class Memory extends Adapter
 	/**
 	 * Checks if a role has access to a resource
 	 */
-	protected function _allowOrDeny(string roleName, string resourceName, var access, var action)
+	protected function _allowOrDeny(string roleName, string resourceName, var access, var action, var func)
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: BC breaks?

File: phalcon/acl/adapter/memory.zep [hunk 9]
```diff
@@ -136,6 +138,13 @@ class Memory extends Adapter
 	protected _accessList;
 
 	/**
+     * Function List
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: indents

File: phalcon/acl/adapter/memory.zep [hunk 10]
```diff
@@ -530,6 +564,10 @@ class Memory extends Adapter
 			let haveAccess = accessList[accessKey];
 		}
 
+		if isset funcList[accessKey] {
+			let funcAccess = funcList[accessKey];
+		}
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: `fetch funcAccess, funcList[accessKey]`

**Comment by Jurigag** [comment]
Also should i check if func is not null before adding it to _func ? Is fetch not setting value if returned value is null ?

```
fetch this->_func[accessKey] = func;
```

Something like this is fine ?

File: phalcon/acl/adapter/memory.zep [hunk 11]
```diff
@@ -495,17 +510,36 @@ class Memory extends Adapter
...
 	{
 		var eventsManager, accessList, accessKey,
 			haveAccess = null, roleInherits, inheritedRole, rolesNames,
-			inheritedRoles;
+			inheritedRoles, funcAccess = null, resourceObject = null, roleObject = null, funcList;
+
+
+
+		if typeof roleName == "object" {
+			if roleName instanceof Roleable {
```
Comment (sergeyklay) [inline phalcon/acl/adapter/memory.zep]: try this:

```zephir
		if typeof roleName == "object" 
				if !(roleName instanceof Roleable) {
						throw new Exception("Object passed as roleName must implement Roleable");
				}
			let roleName = roleName->getName();
		}
```

and below

**Comment by Jurigag** [comment]
Should we notice that there is function if roleObject or resourceObject is null(just passed normally as string) and we found function for accessKey ?

**Comment by Jurigag** [comment]
And maybe add option to add function without any params and just with roleable/resourceable.

**Comment by Jurigag** [comment]
Also i thinking about custom parameters(array) in isAllowed method. Then we can create any function with any parameters in allow method and then just pass them isAllowed method.

Tell me what you think about i too.

**Comment by Jurigag** [comment]
```
CacheTest::testDataFileCacheDecrement
Failed asserting that null matches expected 95.
/home/travis/build/phalcon/cphalcon/unit-tests/CacheTest.php:190
```

;o thats weird i didnt change there anything @sergeyklay

**Comment by Jurigag** [comment]
Added all things i wrote here. We can pass custom parameters as an associative array in isAllowed function where key is parameter name in our function. Also there is method what to do if there function exists but we didnt pass any required parameters to it, just working same as setDefaultAccess.

**Comment by sergeyklay** [comment]
Could you rebase and squash it into a single commit to keep the history clean? Let me know if you need instructions. Also update please `CHANGELOG.md`

File: phalcon/acl/adapterinterface.zep [hunk 12]
```diff
@@ -38,6 +38,16 @@ interface AdapterInterface
...
 
 	/**
+	 * Sets the default access level (Phalcon\Acl::ALLOW or Phalcon\Acl::DENY) for no arguments provided in isAllowed action if there exists func for accessKey
+	 */
+	public function setNoArgumentsDefaultAction(int defaultAccess);
+
+	/**
+	 * Returns the default ACL access level for no arguments provided in isAllowed action if there exists func for accessKey
+	 */
+	public function getNoArgumentsDefaultAction() -> int;
```
Comment (sergeyklay) [inline phalcon/acl/adapterinterface.zep]: These methods BC breaks. In this case this PR should be for 2.1.x branch

Comment (Jurigag) [inline phalcon/acl/adapterinterface.zep]: You mean binary compability ? Cuz i dont know how it breaks backward compability. Well i can just do one commit for 2.1.x i guess.

**Comment by Jurigag** [comment]
Branch 2.1.x dont have this fixed : https://github.com/phalcon/cphalcon/commit/1ac2ed8389e209a816978084e0fa3089a20872fc

Should i do PR with this thing fixed ?

**Comment by sergeyklay** [comment]
No. These changes will be merged from 2.0.x

**Pull request closed without merging**